src/main/java/se/liu/ida/hefquin/engine/queryproc/impl/loptimizer/heuristics/RemoveUnnecessaryL2gAndG2l.java

```diff
@@ -73,36 +72,23 @@ protected boolean checkIfL2gOrG2lNeeded( final LogicalPlan plan ){
 	}
 
 	protected static Set<TriplePattern> extractTPs( final LogicalPlan plan ) {
-		final LogicalOperator lop = plan.getRootOperator();
+		final LogicalOperator rootOp = plan.getRootOperator();
 
-		if( lop instanceof LogicalOpRequest) {
-			return LogicalOpUtils.getTriplePatternsOfReq( (LogicalOpRequest<?, ?>) lop);
+		if( rootOp instanceof LogicalOpRequest) {
+			return LogicalOpUtils.getTriplePatternsOfReq( (LogicalOpRequest<?, ?>) rootOp);
 		}
-		else if ( lop instanceof LogicalOpMultiwayUnion || lop instanceof LogicalOpUnion) {
-			final int numOfSubPlans = plan.numberOfSubPlans();
-			Set<TriplePattern> previousTPs = null;
-
-			for ( int i = 0; i < numOfSubPlans; i++ ) {
-				final LogicalOperator subLop = plan.getSubPlan(i).getRootOperator();
-
-				if ( subLop instanceof LogicalOpRequest ) {
-					final Set<TriplePattern> currentTPs = LogicalOpUtils.getTriplePatternsOfReq( (LogicalOpRequest<?, ?>) subLop);
-					if( !currentTPs.isEmpty() && previousTPs != null && !currentTPs.equals( previousTPs) ) {
-						throw new IllegalArgumentException("UNION is not added as a result of source selection");
-					}
-					previousTPs = currentTPs;
-				}
-				else {
-					throw new IllegalArgumentException("Unsupported type of subquery under UNION (" + subLop.getClass().getName() + ")");
-				}
+		else if ( rootOp instanceof BinaryLogicalOp || rootOp instanceof NaryLogicalOp) {
```

Comment:
It would be better to list the Java class of each operator explicitly. That way, we are sure to throw the IllegalArgumentException whenever we add new types of operators and can then think about whether we need to do something special for them here.